apis/v1beta1/config.go

```diff
@@ -371,24 +374,53 @@ type Service struct {
 	Pipelines map[string]*Pipeline `json:"pipelines" yaml:"pipelines"`
 }
 
-// MetricsPort gets the port number for the metrics endpoint from the collector config if it has been set.
-func (s *Service) MetricsPort() (int32, error) {
+// MetricsEndpoint gets the port number and host address for the metrics endpoint from the collector config if it has been set.
+func (s *Service) MetricsEndpoint() (string, int32, error) {
+	defaultAddr := "0.0.0.0"
 	if s.GetTelemetry() == nil {
 		// telemetry isn't set, use the default
-		return 8888, nil
+		return defaultAddr, 8888, nil
 	}
-	_, port, netErr := net.SplitHostPort(s.GetTelemetry().Metrics.Address)
+	host, port, netErr := net.SplitHostPort(s.GetTelemetry().Metrics.Address)
 	if netErr != nil && strings.Contains(netErr.Error(), "missing port in address") {
-		return 8888, nil
+		return defaultAddr, 8888, nil
 	} else if netErr != nil {
-		return 0, netErr
+		return "", 0, netErr
 	}
 	i64, err := strconv.ParseInt(port, 10, 32)
 	if err != nil {
-		return 0, err
+		return "", 0, err
+	}
+
+	if host != "" {
+		defaultAddr = host
+	}
+
+	return defaultAddr, int32(i64), nil
+}
+
+// ApplyDefaults inserts configuration defaults if it has not been set.
+func (s *Service) ApplyDefaults() error {
+	telemetryAddr, telemetryPort, err := s.MetricsEndpoint()
+	if err != nil {
+		return err
+	}
+	tm := &AnyConfig{
+		Object: map[string]interface{}{
+			"metrics": map[string]interface{}{
+				"address": fmt.Sprintf("%s:%d", telemetryAddr, telemetryPort),
+			},
+		},
 	}
 
-	return int32(i64), nil
+	if s.Telemetry == nil {
+		s.Telemetry = tm
+		return nil
+	}
+	if err := mergo.Merge(s.Telemetry, tm); err != nil {
+		return fmt.Errorf("telemetry config merge failed: %w", err)
+	}
```

Comment:
This only overwrites empty values in `s.Telemetry`, right? Maybe it's worthwhile to add a comment about this? I had to go read mergo's docs to figure it out.

Reply:
Well,it should actually be overwritten. Switching to `MergeWithOverwrite`.

Reply:
Should it? I thought we want to leave the endpoint alone if the user set it on their own.

Reply:
mh... It shouldnt rly matter, right? In case the user provides some default, the defaulting will return the default. Means, an overwrite will only take place, in case when both values are equal. :sweat_smile:

Reply:
I changed it back and added a comment.